File: lab/example/java/src/main/java/com/example/docsearch/ui/Handler.java
package com.example.docsearch.ui;

import com.example.docsearch.service.Indexer;
import com.example.docsearch.service.Searcher;
import com.example.docsearch.service.Searcher.SearchResponse;
import com.example.docsearch.types.Types.SearchRequest;
import java.util.Map;

/**
 * UI-Layer. ADR-0001: importiert service + types; KEIN index, embedding direkt.
 */
public final class Handler {

    /** Antwort des UI-Layers: Status und Körper, ohne echten HTTP-Server. */
    public record HttpResult(int status, Map<String, Object> body) { }

    /** Status und Fehler-Code eines abgebildeten Fehlers. */
    public record ErrorMapping(int status, String code) { }

    private final Searcher searcher;
    private final Indexer indexer;

    public Handler(Searcher searcher) {
        this(searcher, null);
    }

    public Handler(Searcher searcher, Indexer indexer) {
        this.searcher = searcher;
        this.indexer = indexer;
    }

    /**
     * Bildet die Fehler-Codes aus spec/spezifikation.md §4 auf HTTP-Status ab.
     * Unbekannte Fehler sind E099.
     */
    public static ErrorMapping statusFor(RuntimeException exc) {
        if (exc instanceof Indexer.DirectoryNotFoundException) {
            return new ErrorMapping(400, "E001");
        }
        if (exc instanceof Searcher.EmptyQueryException) {
            return new ErrorMapping(400, "E002");
        }
        if (exc instanceof Searcher.EmbeddingUnavailableException) {
            return new ErrorMapping(503, "E003");
        }
        return new ErrorMapping(500, "E099");
    }

    public SearchResponse handleSearch(SearchRequest req) {
        return searcher.search(req);
    }

    /** Suche mit Fehler-Abbildung nach spec/spezifikation.md §4. */
    public HttpResult handleSearchHttp(SearchRequest req) {
        try {
            final SearchResponse resp = searcher.search(req);
            return new HttpResult(200, Map.of("results", resp.results()));
        } catch (RuntimeException e) {
            final ErrorMapping m = statusFor(e);
            return new HttpResult(m.status(), Map.of("error", m.code()));
        }
    }

    /** Reindex (LH-FA-01) mit Fehler-Abbildung nach spec/spezifikation.md §4. */
    public HttpResult handleReindex(String directory) {
        if (indexer == null) {
            return new HttpResult(500, Map.of("error", "E099"));
        }
        try {
            final int n = indexer.reindex(directory);
            return new HttpResult(200, Map.of("indexed_docs", n));
        } catch (RuntimeException e) {
            final ErrorMapping m = statusFor(e);
            return new HttpResult(m.status(), Map.of("error", m.code()));
        }
    }
}

File: lab/example/kotlin/src/main/kotlin/com/example/docsearch/ui/Handler.kt
// UI-Layer. ADR-0001: importiert service + types; KEIN index, embedding direkt.
package com.example.docsearch.ui

import com.example.docsearch.service.DirectoryNotFoundException
import com.example.docsearch.service.EmbeddingUnavailableException
import com.example.docsearch.service.EmptyQueryException
import com.example.docsearch.service.Indexer
import com.example.docsearch.service.SearchResponse
import com.example.docsearch.service.Searcher
import com.example.docsearch.types.SearchRequest

private const val STATUS_OK = 200
private const val STATUS_BAD_REQUEST = 400
private const val STATUS_SERVER_ERROR = 500
private const val STATUS_UNAVAILABLE = 503

/** Antwort des UI-Layers: Status und Körper, ohne echten HTTP-Server. */
data class HttpResult(val status: Int, val body: Map<String, Any>)

/**
 * Bildet die Fehler-Codes aus spec/spezifikation.md §4 auf HTTP-Status ab.
 * Unbekannte Fehler sind E099.
 */
fun statusFor(exc: Throwable): Pair<Int, String> = when (exc) {
    is DirectoryNotFoundException -> STATUS_BAD_REQUEST to "E001"
    is EmptyQueryException -> STATUS_BAD_REQUEST to "E002"
    is EmbeddingUnavailableException -> STATUS_UNAVAILABLE to "E003"
    else -> STATUS_SERVER_ERROR to "E099"
}

class Handler(
    private val searcher: Searcher,
    private val indexer: Indexer? = null,
) {
    fun handleSearch(req: SearchRequest): SearchResponse = searcher.search(req)

    /** Suche mit Fehler-Abbildung nach spec/spezifikation.md §4. */
    fun handleSearchHttp(req: SearchRequest): HttpResult =
        runCatching { searcher.search(req) }.fold(
            onSuccess = { HttpResult(STATUS_OK, mapOf("results" to it.results)) },
            onFailure = { failure(it) },
        )

    /** Reindex (LH-FA-01) mit Fehler-Abbildung nach spec/spezifikation.md §4. */
    fun handleReindex(directory: String): HttpResult {
        val ix = indexer ?: return HttpResult(STATUS_SERVER_ERROR, mapOf("error" to "E099"))
        return runCatching { ix.reindex(directory) }.fold(
            onSuccess = { HttpResult(STATUS_OK, mapOf("indexed_docs" to it)) },
            onFailure = { failure(it) },
        )
    }

    private fun failure(exc: Throwable): HttpResult {
        val (status, code) = statusFor(exc)
        return HttpResult(status, mapOf("error" to code))
    }
}
